services/codecov/codecov.service.js

```diff
@@ -2,37 +2,34 @@
 
 const Joi = require('@hapi/joi')
 const { coveragePercentage } = require('../color-formatters')
-const { BaseJsonService } = require('..')
-
-// https://docs.codecov.io/reference#totals
-// A new repository that's been added but never had any coverage reports
-// uploaded will not have a `commit` object in the response and sometimes
-// the `totals` object can also be missing for the latest commit.
-// Accordingly the schema is a bit relaxed to support those scenarios
-// and then they are handled in the transform and render functions.
-const schema = Joi.object({
-  commit: Joi.object({
-    totals: Joi.object({
-      c: Joi.number().required(),
-    }),
-  }),
-}).required()
+const { BaseSvgScrapingService } = require('..')
 
 const queryParamSchema = Joi.object({
-  token: Joi.string(),
+  token: Joi.string().regex(/^\w{10}$/),
+  // https://docs.codecov.io/docs/flags
+  // Flags must be lowercase, alphanumeric, and not exceed 45 characters
+  flag: Joi.string().regex(/^[a-z0-9_]{1,45}$/),
 }).required()
 
+const schema = Joi.object({
+  message: Joi.string()
+    .regex(/^\d{1,3}%|unknown$/)
+    .required(),
+})
+
+const svgValueMatcher = />(\d{1,3}%|unknown)<\/text><\/g>/
+
 const documentation = `
   <p>
     You may specify a Codecov token to get coverage for a private repository.
   </p>
   <p>
-    See the <a href="https://docs.codecov.io/reference#authorization">Codecov Docs</a>
-    for more information about creating a token.
+    You can find the token for your badge in this url:
+    <code>https://codecov.io/{vcsName}/{user}/{repo}/settings/badge</code>
```

Comment:
Does Codecov use a different token mechanism for their API vs their badges? If so then I'm not sure if we'll be able to bundle the existing, flagless, badges and the flag-based badges into a single service class without breaking existing usage of the badge for private repos

Reply:
> Does Codecov use a different token mechanism for their API vs their badges?

Yes. Codecov provides a 10 character token specifically for their badges.

> If so then I'm not sure if we'll be able to bundle the existing, flagless, badges and the flag-based badges into a single service class without breaking existing usage of the badge for private repos

Yes, that's correct. We can handle this by falling back to scrapping svg when the `flag` query param is passed. But I couldn't find a "clean" way to use both `BaseJsonService` and `BaseSvgScrappingService` together for a specific service.

What do you suggest?

Reply:
In that case we'd have to use two separate classes in that case, one that extends BaseJson for flagless and a new one that extends BaseSvg (this one would need a different route). 

IMO it'd be be ideal if we could have a single implementation like you've done here, but our hand is probably going to be forced to maintain BC for private repo badges.

Curious if any maintainers have any thoughts here?

Reply:
I haven't digested this PR, but here's a quick thought:

If this is a matter of separating out the fetching functionality from the base class, that's something we probably could do. We also use two formats in the F-Droid badge. #2518

If it's trying to do backward compatibility that's a different question (and I'd probably need to study this a bit more).

Reply:
I've re-introduced the previous implementation (with some modifications) and fallback to it with `!flag && token && !badgeTokenPattern.test(token)` condition.

This should be enough to maintain backward-compatibility.